### DISCUSSION TITLE
How to: Transaction ID

### DISCUSSION THREAD

**Discussion by Blackcode** [dkackman/chia-dotnet#50] [Q&A] [answered]
Hello, 
How can I get the TxnHash as on xchscan from the library?

https://xchscan.com/txns/0x82ceceac027a8fc345f8f0bfcd493636a5c0b920121f72d4206fd6ad05f501f6 

Thanks,
Dragos E.

**Comment by dkackman** [comment]
If you have the Json representation of the transaction, it looks like the `TxnHash` above is its `name`. The link for the transaction below would be: https://xchscan.com/txns/0x6efe4a390f6ee36333c18784b9d1be48be3fa00fcf7a45feadd354da9b36a2e0

```json
        {
            "additions": [
                {
                    "amount": 13279172454,
                    "parent_coin_info": "0x7b22ca61b363bde0140e90a6a8222f833b3b32764d16cb6b5dbb6325a9bbc60a",
                    "puzzle_hash": "0x10ea5b5a00a2d5ef9a88e45d34a1268627665e26013ed008a6453bb86d5a31f5"
                }
            ],
            "amount": 13279172454,
            "confirmed": true,
            "confirmed_at_height": 2048218,
            "created_at_time": 1653907970,
            "fee_amount": 0,
            "memos": {},
            "name": "0x6efe4a390f6ee36333c18784b9d1be48be3fa00fcf7a45feadd354da9b36a2e0",
            "removals": [],
            "sent": 0,
            "sent_to": [],
            "spend_bundle": null,
            "to_address": "xch1zr49kksq5t27lx5gu3wnfgfxscnkvh3xqyldqz9xg5amsm26x86sh0kzhl",
            "to_puzzle_hash": "0x10ea5b5a00a2d5ef9a88e45d34a1268627665e26013ed008a6453bb86d5a31f5",
            "trade_id": null,
            "type": 0,
            "wallet_id": 1
        }
```

```csharp
  using var cts = new CancellationTokenSource(1500);
  
  _ = await _theWallet.LogIn(cts.Token);
  var wallet = new Wallet(1, _theWallet);
  
  var transactions = await wallet.GetTransactions(cancellationToken: cts.Token);
  var transaction = transactions.FirstOrDefault();
  Assert.IsNotNull(transaction);

  var link = $"https://xchscan.com/txns/{transaction.Name}";
```

**Reply by Blackcode** [reply]
I do not know why getting the name directly does not work for me. I give you an example of one of my incoming transactions the one I have generated using your code:

`{TransactionRecord { Additions = System.Collections.Generic.List`1[chia.dotnet.Coin], Amount = 26239995995, Confirmed = True, ConfirmedAtHeight = 1971177, CreatedAtTime = 1652467313, FeeAmount = 0, TransactionId = 0x40fedb3f96d5b44f78ba18c4dea5d6d90885419063992c28b28b59adc19e85ee, Name = 0x40fedb3f96d5b44f78ba18c4dea5d6d90885419063992c28b28b59adc19e85ee, Removals = System.Collections.Generic.List`1[chia.dotnet.Coin], Sent = 0, SentTo = System.Collections.Generic.List`1[chia.dotnet.SendPeer], IsInMempool = False, SpendBundle = , ToAddress = xch18fnzv4ex3cemlkz7cw6ntnuzxy5svtsn50l9pe6l5ljgwm5y39vs4njnyy, ToPuzzleHash = 0x3a662657268e33bfd85ec3b535cf823129062e13a3fe50e75fa7e4876e848959, TradeId = , Type = INCOMING_TX, WalletId = 1, CreatedAtDateTime = 5/13/2022 6:41:53 PM }}`

as you can see the Name and the transaction ID is the same.

What Chia client have you used? Maybe it has to do something with that. I am using 1.3.5.

**Reply by Blackcode** [reply]
For converting the name into the TXId I have wrote a quick dirty function which I would like to avoid using.
I pass the values of the parent coin to get an usable TXid: get_coin_info_mojo(transaction.Additions.First().ParentCoinInfo, transaction.Additions.First().PuzzleHash, transaction.Additions.First().Amount); 

```
private string get_coin_info_mojo(string parent_coin_info, string puzzle_hash, ulong amount)
        {
            if (parent_coin_info.IndexOf("0x") == 0)
            {
                parent_coin_info = parent_coin_info.Substring(2);
            }
            if (puzzle_hash.IndexOf("0x") == 0)
            {
                puzzle_hash = puzzle_hash.Substring(2);
            }

            byte[] a = HexToByteArray(parent_coin_info);
            byte[] b = HexToByteArray(puzzle_hash);

            string amountHex = string.Format("{0:X}", amount);
            var byte_count = (Convert.ToString((long)amount, 2).Length + 8) >> 3;

            var joined_amountHex = string.Join("0", new byte[byte_count * 2]) + amountHex;
            amountHex = joined_amountHex.Substring(joined_amountHex.Length - (byte_count * 2));


            if (amountHex.Length % 2 == 1)
            {
                amountHex = '0' + amountHex;
            }

            byte[] c = HexToByteArray(amountHex);

            byte[] d = a.Concat(b).Concat(c).ToArray();

            SHA256 sha256Hash = SHA256.Create();
            byte[] bytes = sha256Hash.ComputeHash(d);

            // Convert byte array to a string   
            StringBuilder builder = new StringBuilder();
            for (int i = 0; i < bytes.Length; i++)
            {
                builder.Append(bytes[i].ToString("x2"));
            }

            return builder.ToString();

            byte[] HexToByteArray(string hexstring)
            {
                var bytes = new List<byte>();

                for (int i = 0; i < hexstring.Length / 2; i++)
                    bytes.Add(Convert.ToByte("" + hexstring[i * 2] + hexstring[i * 2 + 1], 16));

                return bytes.ToArray();
            }
        }
```

**Comment by Blackcode** [comment]
@dkackman do you have any clue why I can't manage to make it work as intended?

**Comment by dkackman** [comment, marked as answer]
I am not sure. It isn't working for me at the moment with a quick test with a recent farming payout transaction, though I swear it worked when I posted my response above!. A couple things:

- You're using mainnet right?
- Have you contacted XCHScan asking what they are using for the txnhash? I can only posit that in some cases the hash is just the name, but in other cases additional fields are hashed in.
- The chia python literally aliases `name` with `transaction_id` and I can;t think of any reason why the would use something else, but it sure looks like they do.

in wallet_rpc_api.py get_transaction which is what I based my response on. Their txnhash must be something subtly different.

```python
        return {
            "transaction": (await self._convert_tx_puzzle_hash(tr)).to_json_dict_convenience(self.service.config),
            "transaction_id": tr.name,
        }
```
